enterprise/backend/src/metabase_enterprise/semantic_search/db/migration.clj

```diff
@@ -0,0 +1,112 @@
+(ns metabase-enterprise.semantic-search.db.migration
+  (:require
+   [honey.sql :as sql]
+   [honey.sql.helpers :as sql.helpers]
+   [metabase-enterprise.semantic-search.db.migration.impl :as semantic.db.migration.impl]
+   [metabase.util.log :as log]
+   [next.jdbc :as jdbc])
+  (:import
+   [java.sql SQLException]))
+
+(def ^:private columns
+  "Migration id is picked by use, hence no serial"
+  [[:version :bigint [:primary-key]]
+   [:finished_at :timestamp [:default [:NOW]]]
+   [:status [:varchar 32]]])
+
+(def migration-table-hsql
+  (-> (sql.helpers/create-table :migration :if-not-exists)
+      (sql.helpers/with-columns columns)))
+
+(defn- migration-table-sql
+  []
+  (sql/format migration-table-hsql))
+
+(defn- ensure-migration-table!
+  [tx]
+  (let [sql (migration-table-sql)]
+    (try
+      (jdbc/execute! tx sql)
+      (catch SQLException e
+        (throw (ex-info "Failed to create migration table."
+                        {:type ::migration-table-creation-failed}
+                        e))))
+    nil))
+
+(defn- db-version
+  [tx]
+  (or (:migration/version
+       (jdbc/execute-one! tx (sql/format {:select [:version]
+                                          :from [:migration]
+                                          :order-by [[:version :desc]]
+                                          :limit 1})))
+      -1))
+
+(defn- write-successful-migration!
+  [tx]
+  (jdbc/execute-one! tx (sql/format (-> (sql.helpers/insert-into :migration)
+                                        (sql.helpers/values [{:version semantic.db.migration.impl/code-version
+                                                              :status "success"}])))))
+
+(defn maybe-migrate-schema!
+  [tx opts]
+  (let [db-version (db-version tx)]
+    (cond
+      (= db-version semantic.db.migration.impl/code-version)
+      (log/info "Migration already performed, skipping.")
```

Comment:
I'm seeing this log on boot from fresh db, whereas I was expecting `(db-version)` to return -1 since the table was just created. Is this the expected behavior?

Reply:
Actually my migration table does have a row in it already even though there are no log lines about `"Starting migration from version %d to %d."`. Maybe I forgot to drop it...

Reply:
Yeah that must have been it. After dropping the migration table and restarting, I now see `"Starting migration from version -1 to 0."`, as expected.

Obviously time for lunch 😂